masonry/src/widgets/canvas.rs

```diff
@@ -0,0 +1,165 @@
+// Copyright 2025 the Xilem Authors and the Druid Authors
+// SPDX-License-Identifier: Apache-2.0
+
+#![warn(missing_docs)]
+
+//! A canvas widget.
+
+use std::sync::Arc;
+
+use accesskit::{Node, Role};
+use smallvec::SmallVec;
+use tracing::{trace_span, Span};
+use vello::kurbo::Size;
+use vello::Scene;
+
+use crate::core::{
+    AccessCtx, AccessEvent, BoxConstraints, EventCtx, LayoutCtx, PaintCtx, PointerEvent, QueryCtx,
+    RegisterCtx, TextEvent, Update, UpdateCtx, Widget, WidgetId, WidgetMut,
+};
+
+/// A widget allowing custom drawing.
+pub struct Canvas {
+    draw: Arc<dyn Fn(&mut Scene, Size) + Send + Sync + 'static>,
+    alt_text: Option<String>,
+}
+
+// --- MARK: BUILDERS ---
+impl Canvas {
+    /// Create a new canvas with the given draw function
+    pub fn new(draw: impl Fn(&mut Scene, Size) + Send + Sync + 'static) -> Self {
+        Self::from_arc(Arc::new(draw))
+    }
+
+    /// Create a new canvas from a function already contained in an [`Arc`].
+    pub fn from_arc(draw: Arc<dyn Fn(&mut Scene, Size) + Send + Sync + 'static>) -> Self {
+        Self {
+            draw,
+            alt_text: None,
+        }
+    }
+
+    /// Set the text that will be used to communicate the meaning of the canvas to
+    /// those using screen readers.
+    ///
+    /// Users are strongly encouraged to set alt text for the canvas.
```

Comment:
I have mixed feelings about that sentence. It's a common piece of advice, but one I'd argue is often misinterpreted as "Apply descriptive alt text everywhere".

When to add non-empty alt text or not could be the subject of an entire article; and there's often situations where adding a description makes a page *less* friendly to screen readers.

Since we don't have such an article written yet, I'd default to something like:

```suggestion
    /// Users are strongly encouraged to set alt text for the canvas.
    /// If possible, the alt-text should succinctly describe what the canvas represents.
    ///
    /// If the canvas is decorative or too hard to describe through text, users should set alt text to `""`.
```

Reply:
Happy to go with what you suggest - I'm not a domain expert.